flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -1008,6 +1008,23 @@ private StreamConfig createJobVertex(Integer streamNodeId, OperatorChainInfo cha
 
         jobVertex.setInvokableClass(streamNode.getJobVertexClass());
 
+        // Find the minimal max parallelism in the chain for operators that have a configured max
+        // parallelism (>0). If no operator has max parallelism configured, returns the default -1.
+        int chainMinimalMaxParallelism =
+                chainInfo.getAllChainedNodes().stream()
+                        .mapToInt(StreamNode::getMaxParallelism)
+                        .filter(maxParallelism -> maxParallelism > 0)
+                        .min()
+                        .orElse(-1);
+
+        jobVertex.setMaxParallelism(chainMinimalMaxParallelism);
+
+        boolean anyParallelismInChainConfigured =
+                chainInfo.getAllChainedNodes().stream()
+                        .anyMatch(StreamNode::isParallelismConfigured);
+
+        jobVertex.setParallelismConfigured(anyParallelismInChainConfigured);
```

Comment:
This is called in `createJobVertex` which is called from `createChain`, which is recursive.

Because `createChain` is recursive, I'm wondering whether `chainInfo` (and its nodes) is "complete" at this point; or there might be more vertices added with `maxParallelism` less then the current?

Reply:
If that was true und the chain incomplete, would that not mean that the current implementation that determines is parallelism is configured would also be wrong?

Reply:
IIUC, `createJobVertex` was only called when a chain has been fully traversed(established).

Reply:
I think the check `currentNodeId.equals(startNodeId)` ensures that `createJobVertex`  it's only called once.